uorb: add section on nested messages (#3562)
PX4/PX4-user_guide: en/middleware/uorb.md

## Adding a New Topic

New uORB topics can be added either within the main PX4/PX4-Autopilot repository, or can be added in an [out-of-tree message definition](../advanced/out_of_tree_modules.md#out-of-tree-uorb-message-definitions).

To add new topics, you need to create a new **.msg** "message definition file" in the `msg/` directory, and add the file name to the `msg/CMakeLists.txt` list.
The new file name should follow the CamelCase convention.

A message definition file can define one or more _topics_, which all have the same fields and structure.
By default a definition maps to a single topic that is named using a snake_case version of the message definition file name (for example, `TopicName.msg` would define a topic `topic_name`).
You can also specify multiple topics to be created by the message definition, which is useful when you need several topics that have the same fields and structure (see [Multi-Topic Messages](#multi-topic-messages) below).

The section [Message Definitions](#message-definitions) below describes the message format.

From the message definitions, the needed C/C++ code is automatically generated.

> To each generated C/C++ struct, a field `uint64_t timestamp` will be added.
> This is used for the logger, so make sure to fill it in when publishing the message.

Reviewer: @GuillaumeLaine I've accepted your feedback (thanks), done further updates, and I'm nearly ready to merge. 

The only thing blocking me is this paragraph. The message definitions all seem to have a field `uint64_t timestamp` defined in them. My thinking is that this is not quite correct - because it implies this timestamp is added magically in the generated header. I think it means this:

```suggestion
All message definitions **must** include the `uint64_t timestamp` field as shown, and this should be filled in when publishing the associated topic(s).
This field is needed in order for the logger to be able to record UORB topics.
```

Is that correct? If you don't, will it break anything? (i.e. is it a "should")?

Note I would actually move this in the Message definitions section, I've just suggested it here for convenience

Author: Oh good catch, I wasn't aware either but it seems you're right:
https://github.com/PX4/PX4-Autopilot/blob/0115a0b3d87262674f86bdbd4771c958b993c399/Tools/msg/px_generate_uorb_topic_files.py#L124-L130

If you don't include that exact field in your message, then PX4 will fail to build

### Multi-Topic Messages

Sometimes it is useful to use the same message definition for multiple topics.
This can be specified at the end of the message using a line prefixed with `# TOPICS `, followed by space-separated topic ids.
For example, the [ActuatorOutputs](../msg_docs/ActuatorOutputs.md) message definition is used to define the topic ids as shown:

```text
# TOPICS actuator_outputs actuator_outputs_sim actuator_outputs_debug
```

### Nested Messages

Message definitions can be nested within other messages to create complex data structures.

To nest a message, simply include the nested message type in the parent message definition. For example, [`PositionSetpoint.msg`](../msg_docs/PositionSetpoint.md) is used as a nested message in the [`PositionSetpointTriplet.msg`](../msg_docs/PositionSetpointTriplet.md) topic message definition.

```text
# Global position setpoint triplet in WGS84 coordinates.
# This are the three next waypoints (or just the next two or one).

uint64 timestamp		# time since system start (microseconds)

PositionSetpoint previous
PositionSetpoint current
PositionSetpoint next
```

> 

Reviewer: @GuillaumeLaine A few lines below it says

> The `listener` command is only available on Pixracer (FMUv4) and Linux / OS X.

Is that still true? How can I check the supported boards?

Author: I don't know, I'm not even sure how to check. I **think** the [CONFIG_SYSTEMCMDS_TOPIC_LISTENER=y](https://github.com/PX4/PX4-Autopilot/blob/8d296a50f9b3cf2842dfb5e49037fbbd3ecbe532/boards/airmind/mindpx-v2/default.px4board#L96) board config enables the `listener` command, but I'm not sure.

If this is case, this config does seem enabled for most boards.

What I can tell is that the line you quoted was written in the docs earlier (4 years ago) than the listener board config was enabled in most boards (3 years ago). Seems possible that docs were not updated after the fact

Reviewer: Nice forensics! Also interesting that the boards after fmuv4 are enabled, but the older ones aren't. I'm going to update this as though we were certain.

For information on adding out-of-tree uORB message definitions, please see [this section](../advanced/out_of_tree_modules.md#out-of-tree-uorb-message-definitions).

To add a new topic, you need to create a new **.msg** file in the `msg/` directory and add the file name to the `msg/CMakeLists.txt` list.
The section [Message Definitions](#message-definitions) below describes the message format.

From the message definitions, the needed C/C++ code is automatically generated.

To each generated C/C++ struct, a field `uint64_t timestamp` will be added.
This is used for the logger, so make sure to fill it in when publishing the message.

> To use the topic in the code, first include the generated header, which will be named using the snake-cased version of the (CamelCase) message definition file name.

Author: ```suggestion
To use the topic in the code, first include the generated header, which will be named using the snake_cased version of the (CamelCase) message definition file name.
```

```

In code your refer to the topic using its id, which by default is also the snake_cased version of the message file name: `ORB_ID(velocity_limits)`.
If needed you can have multiple topics defined from the same message, as described below in [Multi-Topic Messages](#multi-topic-messages).

## Message Definitions

The message definition should start with a descriptive _comment_ (a comment starts with the `#` symbol and goes to the end of the line).
The message will then define one or more fields, which are defined with a _type_, such as `bool`, `uint8`, and `float32`, followed by a _name_.
By convention, each field is followed by a descriptive _comment_, which is any text from the `#` symbol to the end of the line.

For example the [VelocityLimits](../msg_docs/VelocityLimits.md) message definition shown below has a descriptive comment, followed by a number of fields, which each have a comment.

```
# Velocity and yaw rate limits for a multicopter position slow mode only

uint64 timestamp # time since system start (microseconds)

# absolute speeds, NAN means use default limit
float32 horizontal_velocity # [m/s]
float32 vertical_velocity # [m/s]
float32 yaw_rate # [rad/s]
```

> By default this message definition will be compiled to a single topic with an id `velocity_limits`, a direct conversion from the CamelCase name to a snake-case version.

Author: ```suggestion
By default this message definition will be compiled to a single topic with an id `velocity_limits`, a direct conversion from the CamelCase name to a snake_case version.
```

This is the simplest form of a message.
See the existing [`msg`](../msg_docs/index.md) files for other examples of how messages are defined.

### Multi-Topic Messages

Sometimes it is useful to use the same message definition for multiple topics.
This can be specified at the end of the message using a line prefixed with `# TOPICS `, followed by space-separated topic ids.

> For example, the [ActuatorOutputs](../msg_docs/ActuatorOutputs.md) message definition is used to define the topic IDs as shown:

Author: ```suggestion
For example, the [ActuatorOutputs](../msg_docs/ActuatorOutputs.md) message definition is used to define the topic ids as shown:
```
more nitpicking, just to be consistent either using ID or id
